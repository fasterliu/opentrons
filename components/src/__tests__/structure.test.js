// structure components tests
import React from 'react'
import {MemoryRouter} from 'react-router'
import Renderer from 'react-test-renderer'

import {PageTabs, TitleBar, VerticalNavBar, NavButton, SidePanel, FILE} from '..'

describe('TitleBar', () => {
  test('adds an h1 with the title', () => {
    const heading = Renderer.create(
      <TitleBar title='hello' />
    ).root.findByType('h1')

    expect(heading).toBeDefined()
    expect(heading.children).toEqual(['hello'])
  })

  test('adds an optional h2 with the subtitle', () => {
    const heading = Renderer.create(
      <TitleBar title='hello' subtitle='world' />
    ).root.findByType('h2')

    expect(heading).toBeDefined()
    expect(heading.children).toEqual(['world'])
  })

  test('add optional back button', () => {
    const onBackClick = jest.fn()
    const button = Renderer.create(
      <TitleBar title='hello' onBackClick={onBackClick} />
    ).root.findByType('button')

    button.props.onClick()
    expect(onBackClick).toHaveBeenCalled()
  })

  test('renders TitleBar without subtitle correctly', () => {
    const tree = Renderer.create(
      <TitleBar title='foo' />
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('renders TitleBar with subtitle correctly', () => {
    const tree = Renderer.create(
      <TitleBar title='foo' subtitle='bar' />
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('renders TitleBar with back button correctly', () => {
    const tree = Renderer.create(
      <TitleBar title='foo' subtitle='bar' onBackClick={() => {}} />
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })
})

describe('PageTabs', () => {
  test('renders h3 links for each page', () => {
    const pages = [
      {title: 'Page1', href: '/page1', isActive: false, isDisabled: false},
      {title: 'Page2', href: '/page2', isActive: false, isDisabled: false}
    ]

    const root = Renderer.create(
      <MemoryRouter>
        <PageTabs pages={pages} />
      </MemoryRouter>
    ).root

    const links = root.findAllByType('a')
    expect(links).toHaveLength(2)

    links.forEach((link, index) => {
      const {title, href} = pages[index]
      expect(link.props.href).toBe(href)
      expect(link.findByType('h3').children).toEqual([title])
    })
  })

  test('does not create a link if disabled', () => {
    const pages = [
      {title: 'Page1', href: '/page1', isActive: false, isDisabled: true}
    ]

    const notLink = Renderer.create(
      <MemoryRouter>
        <PageTabs pages={pages} />
      </MemoryRouter>
    ).root.findByType('span')

    expect(notLink.findByType('h3').children).toEqual([pages[0].title])
  })

  test('adds active class if active', () => {
    const pages = [
      {title: 'Page1', href: '/page1', isActive: true, isDisabled: false}
    ]

    const link = Renderer.create(
      <MemoryRouter>
        <PageTabs pages={pages} />
      </MemoryRouter>
    ).root.findByType('a')

    expect(link.props.className).toMatch(/active/)
  })

  test('renders PageTabs correctly', () => {
    const pages = [
      {title: 'Page1', href: '/page1', isActive: true, isDisabled: false},
      {title: 'Page2', href: '/page2', isActive: false, isDisabled: true}
    ]

    const tree = Renderer.create(
      <MemoryRouter>
        <PageTabs pages={pages} />
      </MemoryRouter>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })
})

describe('VerticalNavBar', () => {
  test('renders correctly', () => {
    const onClick = () => {}
    const tree = Renderer.create(
      <VerticalNavBar onClick={onClick} className='c'>
        children
      </VerticalNavBar>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })
})

describe('NavButton', () => {
  test('creates a button with props', () => {
    const onClick = jest.fn()
    const button = Renderer.create(
      <NavButton
        onClick={onClick}
        disabled={false}
        isCurrent
        iconName={FILE}
      />
    ).root.findByType('button')

    button.props.onClick()
    expect(button.props.disabled).toBe(false)
    expect(button.props.className).toEqual('button active')
    expect(onClick).toHaveBeenCalled()
  })

  test('adds svg icon to button by name', () => {
    const icon = Renderer.create(
      <NavButton iconName={FILE} />
    ).root.findByType('svg')

    expect(icon).toBeDefined()
  })

  test('renders nav button with icon correctly', () => {
    const tree = Renderer.create(
      <NavButton iconName={FILE} disabled='false' />
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })
})

describe('SidePanel', () => {
  test('renders sidebar with title', () => {
    const heading = Renderer.create(
      <SidePanel title={'title'} />
    ).root.findByType('h2')
    expect(heading).toBeDefined()
    expect(heading.children).toEqual(['title'])
  })

  test('renders close button when onClick is present', () => {
    const onClick = jest.fn()
    const button = Renderer.create(
      <SidePanel title={'title'} onCloseClick={onClick} />
    ).root.findByType('button')

    expect(button).toBeDefined()
    button.props.onClick()
    expect(onClick).toHaveBeenCalled()
  })

  test('renders closed panel when onClick present and isOpen is false', () => {
    const onClick = jest.fn()
    const panel = Renderer.create(
      <SidePanel title={'title'} isClosed='true' onCloseClick={onClick} />
    ).root.findByType('div')

    expect(panel.props.className).toEqual('panel closed')
  })

  test('renders SidePanel correctly', () => {
    const onClick = jest.fn()
    const tree = Renderer.create(
      <SidePanel title={'title'} onCloseClick={onClick} isClosed='true'>
        children
      </SidePanel>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
