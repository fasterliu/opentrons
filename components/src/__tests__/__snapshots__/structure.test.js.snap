// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavButton renders nav button with icon correctly 1`] = `
<button
  className="button"
  disabled="false"
  onClick={undefined}
>
  <svg
    aria-hidden="true"
    className=""
    fill="currentColor"
    height={undefined}
    version="1.1"
    viewBox="0 0 24 24"
    width={undefined}
    x={undefined}
    y={undefined}
  >
    <path
      d="M20.907 0H3.803C3.334.044 3 .356 3 .822c0 1.822.022 3.645.022 5.467v11.355c0 1.823-.022 3.645-.022 5.467 0 .467.334.8.803.822H7.37c2.051 0 3.88.023 6.199 0 .736 0 1.383-.289 1.918-.777.535-.49 1.048-.956 1.538-1.49a201.554 201.554 0 0 0 2.989-3.177c.379-.4.713-.822 1.025-1.267a3.17 3.17 0 0 0 .58-1.844c.045-4.867.045-9.711.067-14.578.022-.444-.312-.778-.78-.8zm-2.074 16.444c-1.115.312-2.275.312-3.434.334-.424 0-.647.155-.736.555-.045.2-.023.867-.023 1.09-.066 1-.044 1.6-.178 2.6-.134.866-.624 1.51-1.762 1.488-2.586-.022-8.005 0-8.161 0V1.556h15.61v13.222c.022 1-.536 1.466-1.316 1.666z"
      fillRule="evenodd"
    />
  </svg>
</button>
`;

exports[`PageTabs renders PageTabs correctly 1`] = `
<nav
  className="page_tabs"
>
  <a
    className="tab_link active_tab_link"
    href="/page1"
    onClick={[Function]}
  >
    <h3
      className="tab_title"
    >
      Page1
    </h3>
  </a>
  <span
    className="tab_link"
    to="/page2"
  >
    <h3
      className="tab_title"
    >
      Page2
    </h3>
  </span>
</nav>
`;

exports[`SidePanel renders SidePanel correctly 1`] = `
<div
  className="panel closed"
>
  <div
    className="title_bar"
  >
    <h2
      className="title"
    >
      title
    </h2>
    <button
      className="button_flat button_icon button_close"
      disabled={undefined}
      onClick={[Function]}
      title="close panel"
    >
      <svg
        aria-hidden="true"
        className="button_only_icon"
        fill="currentColor"
        height={undefined}
        version="1.1"
        viewBox="0 0 24 24"
        width={undefined}
        x={undefined}
        y={undefined}
      >
        <path
          d="M18 7.209L16.791 6 12 10.791 7.209 6 6 7.209 10.791 12 6 16.791 7.209 18 12 13.209 16.791 18 18 16.791 13.209 12z"
          fillRule="evenodd"
        />
      </svg>
    </button>
  </div>
  <div
    className="panel_contents"
  >
    children
  </div>
</div>
`;

exports[`TitleBar renders TitleBar with back button correctly 1`] = `
<header
  className="title_bar"
>
  <button
    className="button_flat back_button"
    disabled={undefined}
    onClick={[Function]}
    title="back"
  >
    <svg
      aria-hidden="true"
      className="icon"
      fill="currentColor"
      height={undefined}
      version="1.1"
      viewBox="0 0 24 24"
      width={undefined}
      x={undefined}
      y={undefined}
    >
      <path
        d="M15.429 17.019l-4.242-4.59 4.242-4.59-1.312-1.41-5.545 6 5.545 6z"
        fillRule="evenodd"
      />
    </svg>
    Back
  </button>
  <div
    className="title_wrapper"
  >
    <h1
      className="title"
    >
      foo
    </h1>
    <span
      className="separator"
    >
      |
    </span>
    <h2
      className="subtitle"
    >
      bar
    </h2>
  </div>
</header>
`;

exports[`TitleBar renders TitleBar with subtitle correctly 1`] = `
<header
  className="title_bar"
>
  <div
    className="title_wrapper"
  >
    <h1
      className="title"
    >
      foo
    </h1>
    <span
      className="separator"
    >
      |
    </span>
    <h2
      className="subtitle"
    >
      bar
    </h2>
  </div>
</header>
`;

exports[`TitleBar renders TitleBar without subtitle correctly 1`] = `
<header
  className="title_bar"
>
  <div
    className="title_wrapper"
  >
    <h1
      className="title"
    >
      foo
    </h1>
  </div>
</header>
`;

exports[`VerticalNavBar renders correctly 1`] = `
<nav
  className="navbar c"
>
  children
</nav>
`;
